<a href="https://colab.research.google.com/github/Locung201/Group3_PTDLUD_HCMUS_2425/blob/main/Group3_PTDLUD.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

## **Nhóm 3 - CSC12110 - PTDLUD**

21127211 Nguyễn Vũ Tường An

21127699 Lô Thủy Tiên

# I. Chuẩn bị Dữ liệu (Data Preparation)
### 1. Tải dữ liệu từ nguồn cung cấp


In [ ]:
#Uploading files from your local file system
from google.colab import files
uploaded = files.upload()

import pandas as pd
data = pd.read_csv("2425_QC.csv", index_col=0)

Saving 2425_QC.csv to 2425_QC (3).csv


###2. Kiểm tra cấu trúc dữ liệu

In [ ]:
print("Number of rows and columns in the data:", data.shape)

print("Information about the data structure:")
data.info()

print("Names of the attributes in the dataset:", data.columns)

print("Some top rows of the data:")
data.head()

Number of rows and columns in the data: (1002, 9)
Information about the data structure:
<class 'pandas.core.frame.DataFrame'>
Index: 1002 entries, 68.95 to 36.91
Data columns (total 9 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   Age                   1002 non-null   float64
 1   Area Income           1002 non-null   float64
 2   Daily Internet Usage  1002 non-null   float64
 3   Ad Topic Line         1002 non-null   object 
 4   City                  1002 non-null   object 
 5   Male                  1002 non-null   int64  
 6   Country               1002 non-null   object 
 7   Timestamp             1002 non-null   object 
 8   Clicked on Ad         1002 non-null   int64  
dtypes: float64(3), int64(2), object(4)
memory usage: 78.3+ KB
Names of the attributes in the dataset: Index(['Age', 'Area Income', 'Daily Internet Usage', 'Ad Topic Line', 'City',
       'Male', 'Country', 'Timestamp', 'Clicked on Ad'],
      d

Age  Area Income  Daily Internet Usage  \
Daily Time Spent on Site                                            
68.95                     35.0     61833.90                256.09   
80.23                     31.0     68441.85                193.77   
69.47                     26.0     59785.94                236.50   
74.15                     29.0     54806.18                245.89   
68.37                     35.0     73889.99                225.58   

                                                  Ad Topic Line  \
Daily Time Spent on Site                                          
68.95                        Cloned 5thgeneration orchestration   
80.23                        Monitored national standardization   
69.47                          Organic bottom-line service-desk   
74.15                     Triple-buffered reciprocal time-frame   
68.37                             Robust logistical utilization   

                                    City  Male     Country        Timestamp  \
Daily Time Spent on Site                                                      
68.95                        Wrightburgh     0     Tunisia   27/3/2016 0:53   
80.23                          West Jodi     1       Nauru    4/4/2016 1:39   
69.47                           Davidton     0  San Marino  13/3/2016 20:35   
74.15                     West Terrifurt     1       Italy   10/1/2016 2:31   
68.37                       South Manuel     0     Iceland    3/6/2016 3:36   

                          Clicked on Ad  
Daily Time Spent on Site                 
68.95                                 0  
80.23                                 0  
69.47                                 0  
74.15                                 0  
68.37                                 0

#II. Khám phá và tiền xử lý dữ liệu (Data Preprocessing)

###1. Handling Missing Value & Empty Data:

Mô tả số lượng missing value của mỗi thuộc tính

In [ ]:
# Kiểm tra các giá trị bị thiếu
print("\nMissing Values:")
missing_values = data.isnull().sum()
print(missing_values)


Missing Values:
Age                     1
Area Income             0
Daily Internet Usage    0
Ad Topic Line           0
City                    0
Male                    0
Country                 0
Timestamp               0
Clicked on Ad           0
dtype: int64


**Nhận xét: Chỉ có 1 giá trị bị thiếu trong cột Age.*

 Xử lý dữ liệu missing : Việc điền giá trị thiếu trong cột "Age" bằng giá trị trung vị giúp đảm bảo rằng dữ liệu không bị thiên lệch do giá trị ngoại lệ và vẫn duy trì tính chính xác cho các phân tích và mô hình dự đoán tiếp theo.


In [ ]:
# Điền giá trị thiếu cho cột 'Age' bằng giá trị trung vị
data['Age'].fillna(data['Age'].median(), inplace=True)

# Kiểm tra lại số lượng giá trị thiếu sau khi điền
print("\nMissing Values After Filling:")
missing_values_after = data.isnull().sum()
print(missing_values_after)


Missing Values After Filling:
Age                     0
Area Income             0
Daily Internet Usage    0
Ad Topic Line           0
City                    0
Male                    0
Country                 0
Timestamp               0
Clicked on Ad           0
dtype: int64


<ipython-input-22-4bac8d6e0cb4>:2: FutureWarning: A value is trying to be set on a copy of a DataFrame or Series through chained assignment using an inplace method.
The behavior will change in pandas 3.0. This inplace method will never work because the intermediate object on which we are setting values always behaves as a copy.

For example, when doing 'df[col].method(value, inplace=True)', try using 'df.method({col: value}, inplace=True)' or df[col] = df[col].method(value) instead, to perform the operation inplace on the original object.


  data['Age'].fillna(data['Age'].median(), inplace=True)


** Nhận xét: Kết quả kiểm tra cho thấy không còn giá trị nào bị thiếu trong tất cả các cột, điều này cho thấy dữ liệu hiện tại đã đầy đủ và sẵn sàng cho các phân tích tiếp theo.*


###3.Handling Duplications | Errors

Mô tả số lượng Duplications | Errors

In [ ]:
# Kiểm tra các dòng trùng lặp
duplicates = data.duplicated()
print("\nDuplicate rows:")
print(duplicates)


Duplicate rows:
Daily Time Spent on Site
68.95    False
80.23    False
69.47    False
74.15    False
68.37    False
         ...  
51.30    False
51.63    False
55.55    False
45.01    False
36.91     True
Length: 1002, dtype: bool


Sử dụng phương thức drop_duplicates(keep='first') để giữ lại dòng đầu tiên của các dòng trùng lặp và xóa các dòng còn lại.

In [ ]:
# Giữ lại dòng đầu tiên và xóa các dòng trùng lặp
data_cleaned_first = data.drop_duplicates(keep='first')

# Hiển thị thông tin về DataFrame sau khi xóa các dòng trùng lặp
print("\nDataFrame after keeping the first occurrence:")
print(data_cleaned_first)

# Kiểm tra lại số lượng dòng trong DataFrame trước và sau khi xóa
print("\nNumber of rows before removal:", len(data))
print("Number of rows after removal:", len(data_cleaned_first))


DataFrame after keeping the first occurrence:
                           Age  Area Income  Daily Internet Usage  \
Daily Time Spent on Site                                            
68.95                     35.0     61833.90                256.09   
80.23                     31.0     68441.85                193.77   
69.47                     26.0     59785.94                236.50   
74.15                     29.0     54806.18                245.89   
68.37                     35.0     73889.99                225.58   
...                        ...          ...                   ...   
72.97                     30.0     71384.57                208.58   
51.30                     45.0     67782.17                134.42   
51.63                     51.0     42415.72                120.37   
55.55                     19.0     41920.79                187.95   
45.01                     35.0     29875.80                178.35   

                                                  Ad To

**Nhận xét: Kết quả cho thấy số lượng dòng đã giảm từ 1002 xuống còn 1000, nghĩa là có 2 dòng trùng lặp đã bị xóa. DataFrame sau khi xóa các dòng trùng lặp vẫn giữ nguyên cấu trúc và các cột, cho thấy rằng không có thông tin nào bị mất ngoài các dòng trùng lặp.*
